python/ml4ir/base/model/calibrations/temperature_scaling.py

```diff
@@ -0,0 +1,247 @@
+"""
+Implementing temperature scaling proposed in https://github.com/gpleiss/temperature_scaling
+The tensorflow implementation and functions are inspired by
+https://www.tensorflow.org/probability/examples/Optimizers_in_TensorFlow_Probability
+"""
+import sys
+import os
+import contextlib
+import functools
+import time
+import tensorflow as tf
+import numpy as np
+import pandas as pd
+import tensorflow_probability as tfp
+from logging import Logger
+from ml4ir.base.data.relevance_dataset import RelevanceDataset
+from ml4ir.base.model.scoring.scoring_model import ScorerBase
+
+import shutil
+
+TEMPERATURE_SCALE = 'temp_scaling_scores'
+
+
+def make_val_and_grad_fn(fn):
+    """ Returns a function that computes y = fn(*args, **kwargs) and gradients of y wrt
+    each of args and  kwargs
+    parameters:
+        fn : Python callable to be differentiated
+    """
+    @functools.wraps(fn)
+    def val_and_grad(x):
+        return tfp.math.value_and_gradient(fn, x)
+    return val_and_grad
+
+
+@contextlib.contextmanager
+def timed_execution(logger):
+    """ a context manager to compute the execution time of a process usually within
+     a 'with' statement
+    Parameter:
+        logger: logging.Logger object
+    """
+    t0 = time.time()
+    yield
+    dt = time.time() - t0
+    logger.info(f'Evaluation took: {dt} seconds')
+
+
+def np_value(tensor):
+    """Get numpy value out of possibly nested tuple of tensors.
+    Parameters:
+        tensor: TF.Tensor
+                a tensor object
+    """
+    if isinstance(tensor, tuple):
+        return type(tensor)(*(np_value(t) for t in tensor))
+    else:
+        return tensor.numpy()
+
+
+def run(optimizer, logger):
```

Comment:
[Minor] @marjanhs I would probably call this `tune_probability` or something. `run` was a little confusing.

Reply:
@lastmansleeping  how about `run_optimizer` ?

Reply:
Works for me.